source/gui/inputGestures.py

```diff
@@ -0,0 +1,799 @@
+			return
+
+		catVM, scriptVM, gestureVM = self.tree.getSelectedItemData()
+		log.debug(f"selection: {catVM}, {scriptVM}, {gestureVM}")
+
+		if scriptVM is None and isinstance(catVM, _EmuCategoryVM):
+			self.gesturesVM.isExpectingNewEmuGesture = catVM
+			pending = catVM.createPendingEmuGesture()
+			self.tree.doRefresh(focus=(catVM, pending, None))
+			self._refreshButtonState()
+
+			def addKbEmuGestureCaptor(gesture):
+				if not isinstance(gesture, keyboardHandler.KeyboardInputGesture) or gesture.isModifier:
+					return False
+				inputCore.manager._captureFunc = None
+				wx.CallAfter(self._addCapturedKbEmu, gesture, catVM)
+				return False
+
+			inputCore.manager._captureFunc = addKbEmuGestureCaptor
+		elif gestureVM is None and isinstance(scriptVM, _ScriptVM):
+			self.gesturesVM.isExpectingNewGesture = scriptVM
+			pendingGesture = scriptVM.createPendingGesture()
+			self.tree.doRefresh(focus=(catVM, scriptVM, pendingGesture))
+			self._refreshButtonState()
+
+			def addGestureCaptor(gesture):
+				if gesture.isModifier:
+					return False
+				if isinstance(catVM, _EmuCategoryVM):
+					gesName = keyLabels.getKeyCombinationLabel(gesture.normalizedIdentifiers[-1][3:])
+					if gesName == scriptVM.scriptInfo.displayName:
+						# Disallow assigning an emulated gesture to itself
+						return False
+				inputCore.manager._captureFunc = None
+				wx.CallAfter(self._addCaptured, catVM, scriptVM, gesture)
+				return False
+
+			inputCore.manager._captureFunc = addGestureCaptor
+		else:
+			log.error(f"unable to do 'add' action for selected item")
+
+	def _addCaptured(self, catVM: _CategoryVMTypes, scriptVM: _ScriptVMTypes, gesture):
+		gids = gesture.normalizedIdentifiers
+		if len(gids) > 1:
+			# Multiple choices. Present them in a pop-up menu.
+			menu = wx.Menu()
+			for gid in gids:
+				gestureVM = _GestureVM(gid)
+				item = menu.Append(wx.ID_ANY, gestureVM.displayName)
+				self.Bind(
+					wx.EVT_MENU,
+					lambda evt, gid=gid: self._addChoice(catVM, scriptVM, gid),
+					item
+				)
+			self.PopupMenu(menu)
+			if self.gesturesVM.isExpectingNewGesture:
+				# No item was selected, so use the first.
+				self._addChoice(catVM, scriptVM, gids[0])
+			menu.Destroy()
+		else:
+			self._addChoice(catVM, scriptVM, gids[0])
+
+	def _addChoice(self, catVM: _CategoryVMTypes, scriptVM: _ScriptVMTypes, gid: str):
+		"""
+		:param scriptVM: The script to add the gesture to
+		:param gid:  Normalized gesture ID to be added.
+		:return:
+		"""
+		newItem = scriptVM.finalisePending(gid)
+		log.debug(f"New: {catVM}, {scriptVM}, {newItem}")
+		self.gesturesVM.isExpectingNewGesture = None
+		self.tree.doRefresh(focus=(catVM, scriptVM, newItem))
+		self._refreshButtonState()
+
+	def _addCapturedKbEmu(self, gesture, catVM: _EmulatedGestureVM):
+		assert isinstance(catVM, _EmuCategoryVM)
+		# Use the last normalized identifier, which is the most generic one
+		gestureToEmulate = gesture.normalizedIdentifiers[-1]

```

Comment:
Ibelieve this should actually be identifiers, not normalizedIdentifiers on this line.
The reason for this is that eventually this string will be passed to KeyboardInputGesture.fromName, but fromName cannot currently cope with a normalized identifier (where the main key may not necessarily appear after the modifiers). 
Take the example of trying to emulate NvDA+f:
identifiers is kb:nvda+f but normalizedIdentifiers is kb:f+nvda (as it has been reordered in a determinate way for internal matching). 
But, if you pass "f+nvda" to KeyboardInputGesture.fromName, you get back a rather broken gesture with normalizedIdentifier of "kb:+nvda".
You could argue this is a bug in KeyboardInputGesture.fromName, but I'm not so sure as fromName really should only handle a user-acceptable representation, and not any kind of internal representation. 
Changing normalizedIdentifiers to identifiers on this line then will correctly emulate kb:nvda+f, correctly matching and executing the reportFormatting script, or if that was not mapped, correctly sending it on to the OS.
Note that this also affects emulating shift+downArrow. In deed anything that gets reordered in normalizedIdentifiers.

Reply:
The thing that worries me about using a non-normalized name as the key for emulations is the potential for bugs based due to localization / or different orderings of the same gesture. But maybe identifiers is still acceptable for this purpose?

Reply:
In particular the docs for `inputCore.InputGesture._get_identifiers`

> If C{id} contains multiple chunks separated by a + sign, they are considered to be ordered arbitrarily
		and may be reordered when normalized.
		Normalization also ensures that the entire identifier is lower case.
		For example, NVDA+control+f1 and control+nvda+f1 will match when normalized.
		See L{normalizeGestureIdentifier} for more details.

Reply:
Only worrying about KeyboardInputGesture (as we only emulate keyboard input) we just need a value that we know will work with KeyboardInputGesture.fromName. It will never be used anywhere else. I can tell you that KeyboardInputGesture.identifiers predictably produces a value that is modifiers+mainKey, which is what fromName expects. Where as normalizeIdentifiers is re-ordered for internal matching of gestures. Remember here the goal is to make a new Gesture, not match against one. 
If you really don't want to use identifiers here, then you'll need to change KeyboardInputGesture.fromName to allow the main key to appear before one or more modifiers. Though I feel that kind of change is actually quite a bit more risky as getting it wrong could affect much more in NVDA.